Refactor triangle-mode

barakman opened the pull request:
Handle issue #650

sklbancor commented on fastlane_bot/modes/base_triangle.py:
```diff
@@ -9,301 +9,83 @@
-    ) -> Tuple[List[str], List[Any]]:
-        """
-        Get combos for triangular arbitrage
+    def assert_supported(self):
```
what is this meant to do? doc string would he helpful

sklbancor commented on fastlane_bot/modes/base_triangle.py:
```diff
@@ -9,301 +9,83 @@
-        Get comprehensive combos for triangular arbitrage
+    @abc.abstractmethod
+    def get_combos(self, flashloan_tokens: List[str], CCm: Any, arb_mode: str) -> Tuple[List[str], List[Any]]:
+        ...
```
doc string? generally in abstract methods it is good to not only define the signature but also the semantics.

barakman replied:
Abstract method removed

sklbancor commented on fastlane_bot/modes/base_triangle.py:
```diff
@@ -9,301 +9,83 @@
-            List of flashloan tokens
-        CCm : object
-            CCm object
+    def find_arbitrage(self, candidates: List[Any] = None, ops: Tuple = None, best_profit: float = 0, profit_src: float = 0) -> Union[List, Tuple]:
```
doc string

barakman replied:
Outside the scope of this PR (same code used prior to it)

barakman replied:
That's right. I'll add doc-strings as soon as i complete this refactoring (which currently just keeps growing).

sklbancor commented on fastlane_bot/modes/base_triangle.py:
```diff
@@ -9,301 +9,83 @@
+                    # Failed to converge
+                    continue
+                trade_instructions_df = r.trade_instructions(O.TIF_DFAGGR)
+                trade_instructions = r.trade_instructions()
```
why do you get the trade instructions in all those different formats? they are equivalent and you should choose one

barakman replied:
Outside the scope of this PR (same code used prior to it)

sklbancor replied:
but are they all used below?

barakman replied:
Annoying indeed.
I was initially aiming for the least amount of changes required, but as this task continues to expand, it is becoming more and more reasonable that we may as well clean up this sort of stuff (which seems to exists in each and everyone of the modes, BTW).

sklbancor commented on fastlane_bot/modes/triangle_multi.py:
```diff
@@ -22,65 +20,141 @@ class ArbitrageFinderTriangleMulti(ArbitrageFinderTriangleBase):
+            ]
+        else:
+            all_base_exchange_curves = CCm.byparams(exchange=self.base_exchange).curves
+            for flt in flashloan_tokens:  # may wish to run this for one flt at a time
```
I don't understand the comment?

barakman replied:
Outside the scope of this PR (same code used prior to it)

sklbancor commented on fastlane_bot/modes/triangle_multi.py:
```diff
@@ -22,65 +20,141 @@ class ArbitrageFinderTriangleMulti(ArbitrageFinderTriangleBase):
+                if tkn0 != tkn1
+            ]
+        else:
+            all_base_exchange_curves = CCm.byparams(exchange=self.base_exchange).curves
```
comments on what you are doing and why you are doing it would be helpful

barakman replied:
Outside the scope of this PR (same code used prior to it)

sklbancor commented on fastlane_bot/modes/triangle_multi.py:
```diff
@@ -22,65 +20,141 @@ class ArbitrageFinderTriangleMulti(ArbitrageFinderTriangleBase):
+                    base_direction_pair = base_exchange_curves[0].pair
+                    base_direction_one = [curve for curve in base_exchange_curves if curve.pair == base_direction_pair]
+                    base_direction_two = [curve for curve in base_exchange_curves if curve.pair != base_direction_pair]
+                    assert len(base_exchange_curves) == len(base_direction_one) + len(base_direction_two)
```
cant really fail, no?

barakman replied:
Outside the scope of this PR (same code used prior to it), but good point regardless, so will remove

sklbancor commented on fastlane_bot/modes/triangle_multi.py:
```diff
@@ -22,65 +20,141 @@ class ArbitrageFinderTriangleMulti(ArbitrageFinderTriangleBase):
+                    base_direction_one = [curve for curve in base_exchange_curves if curve.pair == base_direction_pair]
+                    base_direction_two = [curve for curve in base_exchange_curves if curve.pair != base_direction_pair]
+                    assert len(base_exchange_curves) == len(base_direction_one) + len(base_direction_two)
+                    y_match_curves = CCm.bypairs(
```
y takes tknx and vice versa versa. correct?

platonfloria committed:
Refactor triangle-mode

Minor

Change `assert_supported` exception type from `ValueError` to `AssertionError`

Remove unused return-value

Remove unused code

Refactor single-trianglre mode as well

Remove unneeded assertion

Cosmetic

Fix single-triangle mode

Fix single-triangle mode

Refactor the b3-two-hop-triangle mode as well

Remove the importing of `T`

Remove unused code

Remove single-triangle mode, and fix a bunch of problems

Cleanup `calculate_profit` related code

Fix test 042

Fix  test 042

Remove unused code

Some more general cleanup

Cleanup

Remove unused code

Fix test 063

Cleanup

Cleanup

Fix tests

Cleanup

Cleanup

Fix test 042

Remove single_paiwise and multi_paiwise arb modes (major change)

Fix test 045

Fix test 045

Refactor the pairwise arb-mode classes

Ongoing refactor

Refactor triangle arb-mode classes

Cleanup

Remove file `scan_log_errors.py` from the repository

Cleanup

Remove unused code

Code reuse

Fix functions `create_sort_order` and `sort_key`

Commit:
Refactoring

Commit:
Remove unused code

Commit:
Minor

Commit:
Code reuse

Commit:
Improve base.py

Commit:
Cleanup

Commit:
Remove obsolete files

Commit:
Rename

Commit:
Cleanup

Commit:
Revamp all arb-mode classes and their associated tests

Commit:
Cleanup

Commit:
Cleanup

Commit:
Minor

Commit:
Remove unused code

Commit:
Cleanup

Commit:
Cleanup

Commit:
Cleanup

Commit:
Cleanup

Commit:
Fix test 042

Commit:
handle curve direction triangle complete

Commit:
amend logic triangle curves

Commit:
Cleanup

Commit:
Cleanup

Commit:
Improve logging

Commit:
Improve logging

Commit:
Change arb_opp from a tuple of 5 values to a dictionary of 3 values

Commit:
Cosmetic

Commit:
Improve logging

Commit:
Imoprove logging

Commit:
Remove the `run_data_validator` user-option and all associated code

Commit:
Get rid of the `NoArbAvailable` exception, which can never take place

Commit:
Cleanup

Commit:
Remove unused code

Commit:
Remove the `T` dependency

Commit:
Remove unused code

Commit:
Code reuse

Commit:
Simplify `routerhandler._calc_balancer_out_given_in`

Commit:
Simplify function `routerhandler._get_output_trade_by_source_carbon`

Commit:
Cleanup function `routerhandler._calc_carbon_output`

Commit:
Cleanup

Commit:
Improve logging

Commit:
Improve function `build_pstart`

Commit:
Refactor the input `params` passed to function `optimize`

Commit:
Cosmetic

Commit:
1. Refactor the order encoding/decoding, and unify all duplicated versions of it (utils.py)
2. Simplify overlapping-strategy checking for carbon curves (poolsandtokens.py)
3. Simplify the calculation of carbon output (routerhandler.py)
4. Apply some general cleanup in the async-utils files

Commit:
Fix `routerhandler._calc_carbon_output`

Commit:
Allow the gas token as input to function `routerhandler._calc_carbon_output`

Commit:
Cleanup

Commit:
Add a test for overlapping  strategies

Commit:
Minor

Commit:
Remove old test file

Commit:
bugfix triangle complete when pair not found

Commit:
Minor

Commit:
Minor

Commit:
Use 'int' instead of 'safe_int'

Commit:
Cosmetic

Commit:
In `poolandtokens._carbon_to_cpc_dict`, rely on `utils.encodeOrder` instead of duplicating its code

Commit:
Cosmetic

Commit:
Cosmetic

Commit:
Cleanup

Commit:
Cleanup

Commit:
Unify the code across the arb-mode classes

Commit:
Get rid of `num_format`

Commit:
Cleanup

Commit:
Cleanup

barakman committed:
Merge pull request #675 from bancorprotocol/arb-mode-code-unification

Unify the code across the arb-mode classes

Commit:
Cleanup

Commit:
Simplify `pools_to_add_from_contracts` from a list of tuples of5 elements to a "flat" list of elements

Commit:
Reuse function `ArbitrageFinderBase.calculate_profit` in bot.py

Commit:
Improve logging

Commit:
Fix the setting of `UNI_V2_FORKS` in `network.py`

Lesigh-3100 commented on fastlane_bot/modes/triangle_multi.py:
```diff
@@ -8,79 +8,72 @@
+                tknx = non_flt_carbon_curve.tknx
+
+                carbon_curves = self.CCm.bypairs(f"{tknx}/{tkny}").byparams(exchange="carbon_v1").curves
+                if len(carbon_curves) == 0:
```
The check for carbon curves here should be removed - there is no need to fail this mode if no Carbon Curves exist.

Lesigh-3100 commented on fastlane_bot/modes/triangle_bancor_v3_two_hop.py:
```diff
@@ -8,362 +8,54 @@
-            if str(profit) == "nan":
-                self.ConfigObj.logger.debug("profit is nan, skipping")
+            external_curves = [curve for curve in all_curves if curve.params.get("exchange") not in self.ConfigObj.CARBON_V1_FORKS]
+            if len(external_curves) == 0:
```
This check and the check for carbon_curves above should be combined to:
```
if len(external_curves) == 0 and len(carbon_curves) == 0:
    continue
```

barakman replied:
If you meant `or`, then the second part of this check is already applied a few lines above.
If you meant `and` then please clarify the reason.

Lesigh-3100 replied:
We only want to `continue` if there are no curves to pair with the bancor v3 curves. 

If there are external curves or carbon curves, we want to complete the process.

barakman replied:
Revised, can you please take a look?

Lesigh-3100 commented on fastlane_bot/modes/triangle_bancor_v3_two_hop.py:
```diff
@@ -8,362 +8,54 @@
 
-                    if len(base_direction_two) > 0:
-                        miniverses += [bancor_v3_curve_0 + bancor_v3_curve_1 + base_direction_two]
+            miniverses += [bancor_v3_curve_0 + bancor_v3_curve_1 + carbon_curves]
```
This should be removed as the carbon curves are being added separately above.

barakman replied:
Fixed

Commit:
Remove unused code

Commit:
Improve logging

Commit:
Minor

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Fix `base_triangle.get_params`

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Fix `bot.calculate_profit`

Commit:
Remoe redundant combos in `triangle_bancor_v3_two_hop.py`

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Fix `triangle_bancor_v3_two_hop.get_combos`

Commit:
Cosmetic

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Improve `ArbitrageFinderTriangleBancor3TwoHop.get_combos`

Commit:
Cosmetic

Commit:
Improve `ArbitrageFinderBase`

Commit:
Cosmetic

Commit:
Improve `ArbitrageFinderBase.get_profit`

Commit:
Add explicit `.curves` where needed

Commit:
Fix price sorting in class `ArbitrageFinderBase`

Commit:
Minor

Commit:
Cleanup

Commit:
Fix syntax

Commit:
Improve `ArbitrageFinderTriangleMulti.get_combos`

Commit:
Fix `ArbitrageFinderBase.sort_order`

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Update version from 3.2.1 to 4.0.0

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Cleanup

Commit:
Improve function `remove_unmapped_uniswap_v2_pools`

Commit:
Fix price construction

Commit:
Generate a price-map for the optimizer based on the significance of the exchange which the price is taken from (B2, then B3, then U2, then U3, then all the rest with equal significance).

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Simplify the method for finding a reliable price

Commit:
Fix error in previous commit

Commit:
Cosmetic

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Remove unused code

Commit:
Fix tests

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Fix command-line tests

Commit:
Fix tests

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

Commit:
Merge branch 'develop' of https://github.com/bancorprotocol/fastlane-bot into triangle-mode-refactor

Commit:
Merge branch 'triangle-mode-refactor' of https://github.com/bancorprotocol/fastlane-bot into refactor-encode-decode-orders

barakman committed:
Merge pull request #672 from bancorprotocol/refactor-encode-decode-orders

Refactor the order encoding/decoding

Commit:
Remove commented code

Commit:
Cosmetic

Commit:
Simplify function `_swap_token0_in`

Commit:
Remove unused code

Commit:
Extend the test for function `_carbon_to_cpc_dict`

Commit:
Semantic

Commit:
Remove unused code

Commit:
Remove unuwed code

Commit:
Fix type hint

sklbancor commented on fastlane_bot/bot.py:
```diff
@@ -264,46 +243,27 @@ def _get_deadline(self, block_number) -> int:
-        )
-        return {"finder": finder, "r": finder.find_arbitrage()}
+    def get_arb_finder(self, arb_mode: str, flashloan_tokens, CCm) -> ArbitrageFinderBase:
+        return {
```
not a biggie but I personally would define the dict in a class constant that can be inspected when looking at the class and then reference it here rather than defining it only inline.

This also has the advantage that you can include it in tests, eg you can check that the dict keys of available methods are what you expected they are [removing a key there is probably a semver major version event]

sklbancor commented on fastlane_bot/bot.py:
```diff
@@ -264,46 +243,27 @@ def _get_deadline(self, block_number) -> int:
-        data_validator=False,
+        flashloan_tokens: List[str] = None,
+        CCm: CPCContainer = None,
+        arb_mode: str = None,
```
generally I would probably use enums here

sklbancor commented on fastlane_bot/bot.py:
```diff
@@ -312,55 +272,36 @@ def _run(
+            The flashloan tokens
         CCm: CPCContainer
-            The container.
+            The complete market data container
```
I am not sure of the meaning of "complete" here. Is it the entire market at this stage, or only the curves for the optimization?

sklbancor commented on fastlane_bot/bot.py:
```diff
@@ -312,55 +272,36 @@ def _run(
-                    "[bot._run] Data validation failed. Updating pools and restarting."
-                )
-                return
+        arb_opp = rand_item(list_of_items=arb_opps, num_of_items=randomizer)
```
this is consistent with what we are doing now, but arguably randomisation should happen at the node level because they have the complete picture; we should just return everything we find

Commit:
Cleanup routerhandler.py

Commit:
Cleanup

Commit:
Fix tests

Commit:
Retry arb transaction as long as it throws "max fee per gas less than block base fee"

Commit:
Fix error handling

Commit:
Cleanup

Commit:
Fix call to `connect_network` for Tenderly

Commit:
Fix the creation of the XFai-V0-Core contract

Commit:
Merge branch 'develop' of https://github.com/bancorprotocol/fastlane-bot into triangle-mode-refactor

Commit:
Improve coding and performance in pairwise and triangle arb modes

Commit:
Get rid of the `randomizer` input argument, and attempt all profitable opportunities found on every iteration
